pip_audit/_dependency_source/requirement.py

```diff
@@ -273,20 +224,49 @@ def _recover_files(self, tmp_files: list[IO[str]]) -> None:
                 continue
 
     def _collect_preresolved_deps(
-        self,
-        reqs: Iterator[InstallRequirement],
-    ) -> Iterator[tuple[Requirement, Dependency]]:
+        self, reqs: Iterator[InstallRequirement], require_hashes: bool
+    ) -> Iterator[Dependency]:
         """
         Collect pre-resolved (pinned) dependencies.
         """
+        req_names: set[str] = set()
         for req in reqs:
+            if not req.hash_options and require_hashes:
+                raise RequirementSourceError(
+                    f"requirement {req.name} does not contain a hash {str(req)}"
+                )
+            if req.req is None:
+                # For URL requirements that don't have an egg fragment that lists the
+                # package name and version, `pip-requirements-parser` won't attach a
+                # `Requirement` object to the `InstallRequirement`.
+                #
+                # In this case, we can't audit the dependency so we should signal to the
+                # caller that we're skipping it.
+                yield SkippedDependency(
+                    name=req.requirement_line.line,
+                    skip_reason="could not deduce package/specifier pair from requirement, "
+                    "please specify them with #egg=your_package_name==your_package_version",
```

Comment:
I think we need to correct this advice: `#egg=name==version` isn't `pip`-supported syntax. AFAICT, there's no way to pre-declare the package version for a PEP-508-style URL requirement -- the distribution always has to be fetched.

xref: https://github.com/pypa/pip/pull/11617